pkg/kubectl/cmd/util/suggestion_restmapper.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/kubernetes/pkg/kubectl"
+)
+
+type NoResourceWithSuggestionError struct {
+	*meta.NoResourceMatchError
+
+	Suggestions []string
+}
+
+var _ error = &NoResourceWithSuggestionError{}
+
+// suggestionRESTMapper is a decorator of meta.RESTMapper. If the KindFor,
+// KindsFor, ResourcesFor, ResourceFor returns NoResourceMatchError, it will try
+// to guess resource names based on shorted edit distance.
+type suggestionRESTMapper struct {
+	*resourceDiscover
+
+	delegate     meta.RESTMapper
+	excludeShort bool // exclude short form if set as true
+}
+
+var _ meta.RESTMapper = &suggestionRESTMapper{}
+
+func (m *suggestionRESTMapper) KindFor(resource schema.GroupVersionResource) (schema.GroupVersionKind, error) {
+	result, err := m.delegate.KindFor(resource)
+	return result, m.wrapError(err, resource)
+}
+
+func (m *suggestionRESTMapper) KindsFor(resource schema.GroupVersionResource) ([]schema.GroupVersionKind, error) {
+	result, err := m.delegate.KindsFor(resource)
+	return result, m.wrapError(err, resource)
+}
+
+func (m *suggestionRESTMapper) ResourcesFor(resource schema.GroupVersionResource) ([]schema.GroupVersionResource, error) {
+	result, err := m.delegate.ResourcesFor(resource)
+	return result, m.wrapError(err, resource)
+}
+
+func (m *suggestionRESTMapper) ResourceFor(resource schema.GroupVersionResource) (schema.GroupVersionResource, error) {
+	result, err := m.delegate.ResourceFor(resource)
+	return result, m.wrapError(err, resource)
+}
+
+func (m *suggestionRESTMapper) ResourceSingularizer(resource string) (string, error) {
+	return m.delegate.ResourceSingularizer(schema.GroupVersionResource{Resource: resource}.Resource)
+}
+
+func (m *suggestionRESTMapper) RESTMapping(gk schema.GroupKind, versions ...string) (*meta.RESTMapping, error) {
+	return m.delegate.RESTMapping(gk, versions...)
+}
+
+func (m *suggestionRESTMapper) RESTMappings(gk schema.GroupKind, versions ...string) ([]*meta.RESTMapping, error) {
+	return m.delegate.RESTMappings(gk, versions...)
+}
+
+func (m *suggestionRESTMapper) similarResources(resource schema.GroupVersionResource) []string {
```

Comment:
One of the ways to isolate this functionality will be having an independent function `similarResources` which takes in `[]kubectl.ResourceShortcuts` and a `schema.GroupVersionResource` and returns similar resources, so it assumes ResourceShortcuts are passed to it.
And then we can don't have to extend ShortcutExpander, instead just invoke this new function `similarResources` in the NoResourceMatchError error handing.
I think that will address @liggitt concerns about the tight-coupling of this with shortcutexpander or he might have better suggestion about it.

Reply:
thanks @droot for the review. It makes sense. Now I wrapped the suggestion related functionality in side this `suggestionRESTMapper`, so it's already decoupled with shortcut expander.

Reply:
I think this and other Jordan's comment still isn't addressed.

Reply:
I must mis-understand something here. I re-read @liggitt 's comment. @soltysh can you kindly point out which one it doesn't address? I can only think of wrapping line 87-122 to a separate function? If it's about decouping `ShortcutExpander`, it's already decoupled, both `ShortcutExpander` and `suggestionRESTMapper` are decoration to `RESSTMapper`.  suggestion knows nothing about expander. you can basically enable/disable either suggestion or expander separately with this PR.